src/client/datascience/jupyter/kernels/cellExecutionQueue.ts

```diff
@@ -0,0 +1,165 @@
+    /**
+     * Whether all cells have completed processing or cancelled, or some completed & others cancelled.
+     * Even if this property is true, its possible there is still some cleanup remaining (or some promises waiting to be completed).
+     */
+    public get completed(): boolean {
+        return this.cancelledOrCompletedWithErrors || this.completion.completed;
+    }
+    constructor(
+        public readonly editor: NotebookEditor,
+        private readonly notebookPromise: Promise<INotebook>,
+        private readonly executionFactory: CellExecutionFactory,
+        private readonly isPythonKernelConnection: boolean
+    ) {
+        // If the editor is closed, then just stop handling the UI updates.
+        this.editor.onDidDispose(() => this.cancel(true), this, this.disposables);
+        this.completion.promise.finally(() => this.dispose()).catch(noop);
+    }
+    /**
+     * Queue the cells.
+     */
+    public queueCell(cell: NotebookCell) {
+        const existingCellExecution = this.queueOfCellsToExecute.find((item) => item.cell === cell);
+        if (existingCellExecution) {
+            traceCellMessage(cell, 'Use existing cell execution');
+            return existingCellExecution;
+        }
+        const cellExecution = this.executionFactory.create(cell, this.isPythonKernelConnection);
+        this.queueOfCellsToExecute.push(cellExecution);
+        cellExecution.result.finally(() => this.onCellExecutionCompleted(cellExecution));
+        traceCellMessage(cell, 'User queued cell for execution');
+
+        // Start executing the cells.
+        this.startExecutingCells();
+    }
+    /**
+     * Cancel all cells that have been queued & wait for them to complete.
+     * @param {boolean} [forced=false]
+     * If `true`, then do not wait for cell execution to complete gracefully (just kill it).
+     * This is used when we restart the kernel (either as a result of kernel interrupt or user initiated).
+     * When restarted, the execution needs to stop as jupyter will not send more messages.
+     * Hence `forced=true` is more like a hard kill.
+     */
+    public async cancel(forced?: boolean): Promise<void> {
+        this.cancelledOrCompletedWithErrors = true;
+        traceInfo('Cancel pending cells');
+        // Check all cells
+        await Promise.all(this.queueOfCellsToExecute.map((item) => item.cancel(forced)));
+    }
+    /**
+     * Wait for cells to complete (for for the queue of cells to be processed)
+     * If cells are cancelled, they are not processed, & that too counts as completion.
+     */
+    public async waitForCompletion(cells: NotebookCell[] = []): Promise<void> {
+        if (cells.length === 0) {
+            return this.completion.promise;
+        }
+        const executions = Promise.all(
+            this.queueOfCellsToExecute.filter((item) => cells.includes(item.cell)).map((cell) => cell.result)
+        );
+        await Promise.race([executions, this.completion.promise]);
```

Comment:
Not sure why there's a race here? Is there a scenario where the executions don't finish but somehow we're done?

Reply:
E.g. here's the queue of cells `A B C D E`
Assume user runs cell A, then B, then C, then D, then E in order (one by one).
The KernelExecution calls `queue` & needs to wait until each cell is complete.
So, when cell A is queued KernelExecution waits until thats executed. It will call `waitForCOmpletion(A)`.

But we should not have to wait for ALL cells to complete just to let the caller know that cell `B` has completed.
I.e. cell B can complete even while C, D, E are still in the queue & getting processed.

Reply:
Here's the code form KernelExecution
```typescript
    public async executeCell(.....) {
		...

        const executionQueue = this.getOrCreateCellExecutionQueue(editor, notebookPromise);
        executionQueue.queueCell(cell);
        await executionQueue.waitForCompletion([cell]);
	}
```

Reply:
So shouldn't there be a specific completion promise tied to the input passed in then? Why is it waiting for all executions at all?

Reply:
Yes, i can turn `queueCell` into a promise then.
Initially that's what i did, when it was called `runCell`.
I can leave it as `queueCell(): Promise<void>` if that works. Then one doesn't need to call `waitForCompletion`. 
But its weird that the `queueCell` method is async, one does't expect a queueing operation to be async.

Reply:
Heres the suggestion
```typescript
    public async executeCell(.....) {
		...

        const executionQueue = this.getOrCreateCellExecutionQueue(editor, notebookPromise);
		await executionQueue.queueCell(cell);
	}
```

Reply:
Maybe the name shouldn't be queue then? Just execute? 

Why does kernel execution need to know it's even using a queue? It's just a service it uses to dispatch execution.

Reply:
Its called by IKernel that is async.

Reply:
However I would still have the race condition in a private method, I.e. the Race doesn't go away.
Is that ok?

Reply:
Not sure why we have the race still? Seems like the executions complete or they don't? Sorry maybe you explained this above. Let me look some more.

Reply:
So the executions is the cell executions that were asked to run. That makes sense. The caller expects to wait for those. I still don't get why completionPromise is needed? When will that resolve before the execution promises?

Reply:
Only one place, when executing a cell fails.
I.e. if `executeQueueCells` fails for any reason (kernel error, notebook doesn't start, etc...). In such cases the cells complete without errors, but the error needs to be propagated.

Reply:
On second thought, i don't think we do anything with those errors, let me check.
We just need to know whether the cells completed or not (with or without errors).
Give me a moment to double check

Reply:
Yes, removed the promise.
Let me know what you think about the wait now...

Reply:
Thanks, getting a lot cleaner now